src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs

```diff
@@ -0,0 +1,790 @@
+        private static ReadOnlySpan<byte> ArgumentsPropertyNameBytes => new byte[] { (byte)'a', (byte)'r', (byte)'g', (byte)'u', (byte)'m', (byte)'e', (byte)'n', (byte)'t', (byte)'s' };
+        private const string HeadersPropertyName = "headers";
+        private static ReadOnlySpan<byte> HeadersPropertyNameBytes => new byte[] { (byte)'h', (byte)'e', (byte)'a', (byte)'d', (byte)'e', (byte)'r', (byte)'s' };
+
+        private static readonly string ProtocolName = "json";
+        private static readonly int ProtocolVersion = 1;
+        private static readonly int ProtocolMinorVersion = 0;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="JsonHubProtocol"/> class.
+        /// </summary>
+        public JsonHubProtocol() : this(Options.Create(new JsonHubProtocolOptions()))
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="JsonHubProtocol"/> class.
+        /// </summary>
+        /// <param name="options">The options used to initialize the protocol.</param>
+        public JsonHubProtocol(IOptions<JsonHubProtocolOptions> options)
+        {
+        }
+
+        /// <inheritdoc />
+        public string Name => ProtocolName;
+
+        /// <inheritdoc />
+        public int Version => ProtocolVersion;
+
+        /// <inheritdoc />        
+        public int MinorVersion => ProtocolMinorVersion;
+
+        /// <inheritdoc />
+        public TransferFormat TransferFormat => TransferFormat.Text;
+
+        /// <inheritdoc />
+        public bool IsVersionSupported(int version)
+        {
+            return version == Version;
+        }
+
+        /// <inheritdoc />
+        public bool TryParseMessage(ref ReadOnlySequence<byte> input, IInvocationBinder binder, out HubMessage message)
+        {
+            if (!TextMessageParser.TryParseMessage(ref input, out var payload))
+            {
+                message = null;
+                return false;
+            }
+
+            message = ParseMessage(payload, binder);
+
+            return message != null;
+        }
+
+        /// <inheritdoc />
+        public void WriteMessage(HubMessage message, IBufferWriter<byte> output)
+        {
+            WriteMessageCore(message, output);
+            TextMessageFormatter.WriteRecordSeparator(output);
+        }
+
+        /// <inheritdoc />
+        public ReadOnlyMemory<byte> GetMessageBytes(HubMessage message)
+        {
+            return HubProtocolExtensions.GetMessageBytes(this, message);
+        }
+
+        private HubMessage ParseMessage(ReadOnlySequence<byte> input, IInvocationBinder binder)
+        {
+            try
+            {
+                // We parse using the Utf8JsonReader directly but this has a problem. Some of our properties are dependent on other properties
+                // and since reading the json might be unordered, we need to store the parsed content as JsonDocument to re-parse when true types are known.
```

Comment:
Have you considered building a document of the whole payload up front instead of using the reader or is that too expensive? I am asking especially since I see list being used to generate the stream ids.

Reply:
What would be the benefit of making a document? For the most part, we don't hit the JsonDocument code path in normal scenarios since our clients send the data in the "right" order. We need to keep the arguments and return type "un parsed" for as long as possible.

Reply:
The main reason I mentioned this is that you wouldn't have to build up the list and then call ToArray on it. Instead, you could initialize the array based on `element.GetArrayLength()`:
https://github.com/aspnet/AspNetCore/pull/8932/files#diff-81f0f21fef0bc37dc6a70b52a0c3379aR167

But if that is a small part of the payload (or if that list is generally small), it's not worth it.